thorlcr/activities/trace/thtraceslave.cpp

```diff
@@ -0,0 +1,159 @@
+          meta(NULL), keepLimit(0), skip(0), sample(0), traceEnabled(false)
+    {
+        helper = (IHThorTraceArg *) queryHelper();
+        stepCompare = NULL;
+    }
+    void init(MemoryBuffer &data, MemoryBuffer &slaveData)
+    {
+        helper = (IHThorTraceArg *)queryHelper();
+        appendOutputLinked(this);
+    }
+    void start()
+    {
+        ActivityTimer s(totalCycles, timeActivities);
+        dataLinkStart();
+        input = inputs.item(0);
+        startInput(input);
+        IInputSteppingMeta *stepMeta = input->querySteppingMeta();
+        if (stepMeta)
+            stepCompare = stepMeta->queryCompare();
+        meta = helper->queryOutputMeta();
+        traceEnabled = getOptBool(THOROPT_TRACE_ENABLED, false);
+        keepLimit = helper->getKeepLimit();
+        if (keepLimit==(unsigned) -1)
+            keepLimit = getOptUInt(THOROPT_TRACE_LIMIT, 100);
+        if (!traceEnabled)
+            keepLimit = 0;
+        if (!helper->canMatchAny())
+            traceEnabled = false;
+        skip = helper->getSkip();
+        sample = helper->getSample();
+        if (sample)
+            sample--;
+        name.setown(helper->getName());
+        if (!name)
+            name.set("Row");
+    }
+    void stop()
+    {
+        name.clear();
+        stopInput(input);
+        dataLinkStop();
+    }
+    void onTrace(const void *row)
+    {
+        if (keepLimit && helper->isValid(row) && meta)
+        {
+            if (skip)
+                skip--;
+            else if (sample)
+                sample--;
+            else
+            {
+                CommonXmlWriter xmlwrite(XWFnoindent);
+                meta->toXML((const byte *) row, xmlwrite);
+                ActPrintLog("TRACE: <%s>%s<%s>", name.get(), xmlwrite.str(), name.get());
+                keepLimit--;
+                sample = helper->getSample();
+                if (sample)
+                    sample--;
+            }
+        }
+    }
+    CATCH_NEXTROW()
+    {
+        ActivityTimer t(totalCycles, timeActivities);
+        OwnedConstThorRow ret = input->nextRow();
+        if (ret)
+        {
+            onTrace(ret);
+            dataLinkIncrement();
+        }
+        return ret.getClear();
+    }
+    const void *nextRowGE(const void *seek, unsigned numFields, bool &wasCompleteMatch, const SmartStepExtra &stepExtra)
+    {
+        try { return nextRowGENoCatch(seek, numFields, wasCompleteMatch, stepExtra); }
+        CATCH_NEXTROWX_CATCH;
+    }
+    const void *nextRowGENoCatch(const void *seek, unsigned numFields, bool &wasCompleteMatch, const SmartStepExtra &stepExtra)
+    {
+        ActivityTimer t(totalCycles, timeActivities);
+        OwnedConstThorRow ret = input->nextRowGE(seek, numFields, wasCompleteMatch, stepExtra);
+        if (!ret)
+        {
+            abortSoon = true;
+            return NULL;
+        }
+        if (!wasCompleteMatch)
+            return ret.getClear();
+        onTrace(ret);
```

Comment:
I don't know. Debatable. Is a row returned to smart stepper as an incomplete match something we should consider traceable or not? We don't include them in the counts of rows returned, I think (though I must admit I don't really fully understand the smart step code around here)